### PR TITLE
Improve multiparser handle `finalNewLine` and `frontMatter`

### PR DESCRIPTION
**Pull request by fisker**
### changes:

We decided to not accept #8582, making some changes to handle `finalNewLine` easier in `embed`

- Call `stripTrailingHardline()` in `multiparser.js`, so we don't need do this in every `embed` functions
- ~`stripTrailingHardline` function add string support~ reverted, breaking change
- Add `printFrontMatter` to handle all frontMatter print
- Refactor a plugin add in #8465, easier to create a plugin, maybe use it for plugins in `tests_integration\plugins`

I didn't realize `stripTrailingHardline` in `multiparser.js` is a breaking change at first.
I put it behind an option for now, better idea?

### check:
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/document/doc-utils.js [hunk 1]
```diff
@@ -192,6 +192,10 @@ function getInnerParts(doc) {
 }
 
 function stripTrailingHardline(doc, withInnerParts = false) {
+  if (typeof doc === "string") {
+    return doc.replace(/(?:\r?\n)*$/, "");
+  }
```
Comment (evilebottnawi) [inline src/document/doc-utils.js:197]: Breaking change for plugin developers

Comment (fisker) [reply src/document/doc-utils.js:197]: I'll can this in `multiparser.js`, when that option is true

File: src/language-js/embed.js [hunk 2]
```diff
@@ -578,12 +590,16 @@ function printHtmlTemplateLiteral(path, print, textToDoc, parser, options) {
 
   const contentDoc = mapDoc(
     stripTrailingHardline(
```
Comment (sosukesuzuki) [inline src/language-js/embed.js:592]: Can we remove this `stripTrailingHardline` call?

Comment (fisker) [reply src/language-js/embed.js]: Found another one, and removed, end up we only use `stripTrailingHardline` for `embed`.

File: src/document/doc-utils.js (hunk 1 above)
Comment (fisker) [reply src/document/doc-utils.js]: @sosukesuzuki `doc.utils` is not public api? I remember mentioned when you add `normalizeParts`?

Comment (sosukesuzuki) [reply src/document/doc-utils.js]: `doc.utils` is public api.

Comment (fisker) [reply src/document/doc-utils.js]: No? https://github.com/prettier/prettier/pull/8476#discussion_r434645944 , I can't find docs about it.

@evilebottnawi @thorn0

**Comment by fisker** [comment]
I'm going to leave `stripTrailingHardline` for string in `textToDoc` function, and added a `TODO`, maybe we can move it to `stripTrailingHardline` function in next marjor release

**Pull request merged**